### PR TITLE
Adding fingerprints behind flag first cut

### PR DESCRIPTION
**Pull request by microbit-grace**
Also, added view options to data samples menu.
First step to making it work. Design discussions and bits to follow.

### CODE REVIEW

File: lang/ui.en.json [hunk 1]
```diff
@@ -503,6 +519,102 @@
...
+    "description": "Tooltip for data feature of a recording"
+  },
+  "fingerprint-zcr-y-tooltip": {
+    "defaultMessage": "Rate at which acceleration transitions between positive and negative for y-direction data points in {action}",
+    "description": "Tooltip for data feature of a recording"
+  },
+  "fingerprint-zcr-z-tooltip": {
+    "defaultMessage": "Rate at which acceleration transitions between positive and negative for z-direction data points in {action}",
+    "description": "Tooltip for data feature of a recording"
+  },
```
Comment (microbit-grace) [inline lang/ui.en.json:617]: Copy inspired by [https://github.com/microbit-foundation/cctd-ml-machine/blob/a90b7f1f1351d42f8958fc4d932b4201f283ef8a](https://github.com/microbit-foundation/cctd-ml-machine/blob/a90b7f1f1351d42f8958fc4d932b4201f283ef8a/src/messages/ui.en.json#L61-L76)[...](https://github.com/microbit-foundation/cctd-ml-machine/blob/a90b7f1f1351d42f8958fc4d932b4201f283ef8a/src/messages/ui.en.json#L61-L76)

File: src/components/DataRecordingGridItem.tsx [hunk 2]
```diff
@@ -94,13 +98,29 @@ const DataRecordingGridItem = ({
...
+                )}
+                {flags.fingerprints &&
+                  (view === DataSamplesView.DataFeatures ||
+                    view === DataSamplesView.GraphAndDataFeatures) && (
+                    <RecordingFingerprint
+                      data={recording.data}
+                      role="image"
+                      gestureName={data.name}
+                      aria-label={intl.formatMessage({
+                        id: "recording-fingerprint-label",
```
Comment (microbit-matt-hillsdon) [inline src/components/DataRecordingGridItem.tsx:120]: Understandable as there isn't a clear way to represent this data non-visually, at least for the purposes of easy comparison as this feature intends.

Please can you add this to your screenreader notes so we can pick up and consider an alternative?

File: src/mlFilters.ts [hunk 3]
```diff
@@ -78,16 +78,21 @@ const zeroCrossingRate: FilterStrategy = (data) => {
...
+export const mlFilters: Record<
+  Filter,
+  { strategy: FilterStrategy; min: number; max: number }
+> = {
+  [Filter.MAX]: { strategy: (d) => Math.max(...d), min: -2.4, max: 2.4 },
+  [Filter.MIN]: { strategy: (d) => Math.min(...d), min: -2.4, max: 2.4 },
+  [Filter.MEAN]: { strategy: mean, min: -2.4, max: 2.4 },
+  [Filter.STD]: { strategy: stddev, min: 0, max: 2.4 },
+  [Filter.PEAKS]: { strategy: peaks, min: 0, max: 10 },
+  [Filter.ACC]: { strategy: acc, min: 0, max: 160 },
```
Comment (microbit-matt-hillsdon) [inline src/mlFilters.ts:95]: How is 160 arrived at?

Comment (microbit-grace) [reply src/mlFilters.ts:95]: The values for these are from here --> https://github.com/microbit-foundation/cctd-ml-machine/blob/ab6fda7795131e1df865af055d6e0808d1c3000c/src/script/utils/FilterLimits.ts#L27

File: src/components/DataRecordingGridItem.tsx (hunk 2 above)
Comment (microbit-grace) [reply src/components/DataRecordingGridItem.tsx:120]: Makes sense, added it

File: src/mlFilters.ts [hunk 4]
```diff
@@ -78,16 +78,21 @@ const zeroCrossingRate: FilterStrategy = (data) => {
...
+  [Filter.ZCR]: { strategy: zeroCrossingRate, min: 0, max: 1 },
+  [Filter.RMS]: { strategy: rms, min: 0, max: 2 },
```
Comment (microbit-matt-hillsdon) [inline src/mlFilters.ts:97]: Should this be 2.4? Are there practical considerations here?

Comment (microbit-grace) [reply src/mlFilters.ts:97]: Current rms is calculated as such

> const rms: FilterStrategy = (d) =>
>   Math.sqrt(d.reduce((a, b) => a + Math.pow(b, 2), 0) / d.length);

so if max acceleration value is 2 and we assume 80 samples (as Rob mentioned above),  I think max rms should be 2

File: src/mlFilters.ts (hunk 3 above)
Comment (microbit-matt-hillsdon) [reply src/mlFilters.ts:95]: Discussed in chat, we think 2.04 is the actual max and the 160 should be calculated from it.

Comment (microbit-grace) [reply src/mlFilters.ts]: The ml filter limits are updated here: https://github.com/microbit-foundation/ml-trainer/pull/403/commits/edeb739fa06341a7b1ecc0d8d943ac7b9be071e3

File: src/mlFilters.ts (hunk 4 above)
Comment (microbit-grace) [reply src/mlFilters.ts]: The ml filter limits are updated here: https://github.com/microbit-foundation/ml-trainer/pull/403/commits/edeb739fa06341a7b1ecc0d8d943ac7b9be071e3

File: src/utils/gradient-calculator.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+interface RGB {
+  r: number;
+  g: number;
+  b: number;
+}
+
+export const calculateColor = (
```
[the remark this answers is not in the document]
Comment (microbit-matt-hillsdon) [reply src/utils/gradient-calculator.ts:7]: We'll look at this separately from this PR.

**Pull request merged**